### PR TITLE
Fix ca crash (attempt 2)

### LINKED ISSUE #5109: OpenSSL "ca" CLI segfaults on erroneous index.txt
[openssl_ca_segfault.tar.gz](https://github.com/openssl/openssl/files/1645927/openssl_ca_segfault.tar.gz)
I'm trying to setup a minimalistic demo CA using a custom config and somehow ran into a case in which OpenSSL segfaults. See attached tar.gz that gives all necessary prerequisites. Reproducible also with OpenSSL v1.0.2, but also against e44c7d02.

In the files, a demo CA is included using secp521r1 certificates. The CA's index.txt is an empty file (0 bytes). When run for the first time, it works:

```
$ ./sign 
Using configuration from config
Check that the request matches the signature
Signature ok
The Subject's Distinguished Name is as follows
commonName            :ASN.1 12:'bar'
Certificate is to be certified until Jan 17 09:21:31 2028 GMT (3650 days)

Write out database with 1 new entries
Data Base Updated
```

But upon a subsequent run:

```
$ ./sign
Using configuration from config
Check that the request matches the signature
Signature ok
The Subject's Distinguished Name is as follows
commonName            :ASN.1 12:'bar'
./sign: line 4: 23951 Segmentation fault      (core dumped) openssl ca -batch -config config -in csr -subj /CN=bar -out clients/bar.crt -days 3650
```

Stacktrace:

```
Program received signal SIGSEGV, Segmentation fault.
__strcmp_ssse3 () at ../sysdeps/x86_64/multiarch/../strcmp.S:174
174	../sysdeps/x86_64/multiarch/../strcmp.S: No such file or directory.
(gdb) bt
#0  __strcmp_ssse3 () at ../sysdeps/x86_64/multiarch/../strcmp.S:174
#1  0x0000555555587bce in index_name_cmp (a=0x5555558412f0, b=0x7fffffffc900) at apps/apps.c:1357
#2  0x0000555555587c7c in index_name_LHASH_COMP (arg1=0x5555558412f0, arg2=0x7fffffffc900) at apps/apps.c:1363
#3  0x00007ffff77bcac8 in getrn (lh=0x55555583be50, data=0x7fffffffc900, rhash=0x7fffffffc700) at crypto/lhash/lhash.c:308
#4  0x00007ffff77bc471 in OPENSSL_LH_retrieve (lh=0x55555583be50, data=0x7fffffffc900) at crypto/lhash/lhash.c:154
#5  0x00007ffff7840823 in lh_OPENSSL_STRING_retrieve (lh=0x55555583be50, d=0x7fffffffc900) at include/openssl/lhash.h:197
#6  0x00007ffff7840e79 in TXT_DB_get_by_index (db=0x555555842800, idx=5, value=0x7fffffffc900) at crypto/txt_db/txt_db.c:145
#7  0x000055555558f791 in do_body (xret=0x7fffffffcbd0, pkey=0x55555583e3b0, x509=0x555555841e20, dgst=0x7ffff7b08c40 <sha256_md>, sigopts=0x0, policy=0x555555840aa0, 
    db=0x555555842300, serial=0x555555842170, subj=0x7fffffffe831 "/CN=bar", chtype=4097, multirdn=0, email_dn=1, startdate=0x5555555e37fe "today", enddate=0x0, days=3650, 
    batch=1, verbose=0, req=0x5555558423c0, ext_sect=0x55555583e5a0 "x509_exts", lconf=0x55555583f0d0, certopt=0, nameopt=8520479, default_op=1, ext_copy=0, selfsign=0)
    at apps/ca.c:1610
#8  0x000055555558eae7 in certify (xret=0x7fffffffcbd0, infile=0x7fffffffe827 "csr", pkey=0x55555583e3b0, x509=0x555555841e20, dgst=0x7ffff7b08c40 <sha256_md>, sigopts=0x0, 
    policy=0x555555840aa0, db=0x555555842300, serial=0x555555842170, subj=0x7fffffffe831 "/CN=bar", chtype=4097, multirdn=0, email_dn=1, startdate=0x5555555e37fe "today", 
    enddate=0x0, days=3650, batch=1, ext_sect=0x55555583e5a0 "x509_exts", lconf=0x55555583f0d0, verbose=0, certopt=0, nameopt=8520479, default_op=1, ext_copy=0, selfsign=0)
    at apps/ca.c:1300
#9  0x000055555558d5c7 in ca_main (argc=0, argv=0x7fffffffe470) at apps/ca.c:908
#10 0x00005555555a5209 in do_cmd (prog=0x555555830510, argc=12, argv=0x7fffffffe410) at apps/openssl.c:562
#11 0x00005555555a45d0 in main (argc=12, argv=0x7fffffffe410) at apps/openssl.c:181
```

Please note that I think it is *extremely* likely that I'm doing something odd, weird or wrong in my config file. However, I cannot seem to figure out what it is and the SEGV isn't helping me debug the cause.

### PR DESCRIPTION
**Pull request by mattcaswell**
This is another attempt at fixing the issue originally described in #5109. A fix for that was committed as a result of #5114 (#5115 for 1.0.2). Unfortunately that fix prevented empty Subjects from being used completely. As pointed out by @vdukhovni in #5115, an empty Subject is actually permissible in some situations.

This PR reverts the original commits and implements a new fix. The core issue is the name comparison function which cannot handle NULL as an input parameter. This PR updates that function to be able to correctly process that.

Note: that the default database configuration setting is to disallow multiple subjects with the same value. If two certs are issued without a subject then this counts as two with the same value. Therefore to allow this the database should be configured with the "unique_subject = no" setting (see the ca man page for details).

Fixes #5109 

This PR is for master/1.1.0. A separate PR will fix 1.0.2.

### CODE REVIEW

**Comment by vdukhovni** [comment]
Looks good, but I have a question/suggestion. Since an empty Subject DN is semantically "unspecified", as opposed to actually having an empty name, perhaps the "unique_subject" code should allow multiple empty values, and still be able to enforce uniqueness otherwise?

That said, I personally tend to find "unique_subject" a nuisance, and turn it off, so I am not sure why people want "unique_subject".  Perhaps those people would not want the multiple unspecified names?

Finally how are such names recorded in the index file?  As "/", or as an empty string?  The latter might be better if not already the case.

**Comment by levitte** [comment]
`unique_subject` should probably have been set to default to "no" at some point, for example for 1.1.0.  Unfortunately, it was forgotten...

**Comment by vdukhovni** [comment]
Yes, perhaps an unfortunate default, but what do you think about empty (unspecified) subject DNs vs. `unique_subject=yes`?  Should exactly one empty DN be allowed, or should empty DNs each be considered unique, allowing many?

**Comment by mattcaswell** [comment]
> unique_subject should probably have been set to default to "no" at some point, for example for 1.1.0. Unfortunately, it was forgotten...

The ca man page actually recommends that you should change the default. I have created issue #5451 so we don't forget it when we get to 1.2.0.

> Should exactly one empty DN be allowed, or should empty DNs each be considered unique, allowing many?

I kind of think that behaviour might be quite surprising if you were expecting enforcement of no duplicate Subjects.

**Comment by vdukhovni** [comment]
Yeah, perhaps some might be surprised by multiple empty subjects, but they should quickly realize that these are not in fact "duplicates", as the subject name is unspecified, rather than being the unique subject whose name is empty.  It's a judgement call, I am leaving it to the others to make the final decision.  Personally, I would allow empty "duplicates" while enforcing uniqueness otherwise (when enabled/not disabled).

**Comment by mattcaswell** [comment]
@levitte, you have approved this - are you happy for me to continue to merge this in light of @vdukhovni's comments?

**Comment by levitte** [comment]
Hmmmm...

I'm thinking that allowing empty subjects is something new anyway, so we really should make an effort to treat them right from the start, even if that diverges from how we treat other subjects. 
The trick, then, is to know if unique_subject = "yes" is a default (that we allow ourselves to break for empty subjects) or an explicit configuration value (we mustn't break that)

**Comment by mattcaswell** [comment]
> The trick, then, is to know if unique_subject = "yes" is a default (that we allow ourselves to break for empty subjects) or an explicit configuration value (we mustn't break that)

Once the database has been created the index.txt.attr file gets created, populated with the value of unique_subject. From that point on it is impossible to know whether it was set explicitly or by default.

**Comment by levitte** [comment]
Put differently, let's make "no" the default unique_subject for empty subjects now.

**Comment by levitte** [comment]
In that case, I suggest unique_empty_subject with its own default. Or ignore unique_subject for empty subjects.

**Comment by vdukhovni** [comment]
Another way to justify "non-unique" empty subjects is that these are quite likely to actually be certificates for distinct subjects ("principals"), that are identified by the subjectAltName.  It is just the the subjectDN is no longer adequate to distinguish the different principals to which the certificates are issued.

So if Richard agrees, I think we have enough support for this reasonable interpretation, I don't think it requires a team vote, just more evidence of support than my solitary perspective.

**Comment by mattcaswell** [comment]
> Another way to justify "non-unique" empty subjects is that these are quite likely to actually be certificates for distinct subjects ("principals"), that are identified by the subjectAltName. It is just the the subjectDN is no longer adequate to distinguish the different principals to which the certificates are issued.So if Richard agrees, I think we have enough support for this reasonable interpretation, I don't think it requires a team vote, just more evidence of support than my solitary perspective.

Ok. I can live with that.

**Comment by mattcaswell** [comment]
Having empty subjects with unique_empty_subject set to "yes" turns out to be a hard problem to solve. The problem is that we use libcrypto to manage the database. As part of that, if we are going to have unique empty subjects, then we create an index of those subjects. Under pinning the database is a hash, which needs to have a comparison function for elements in the hash. To avoid two different database rows coming back as the same thing if they both have empty subjects, then the comparison function must report them as different. That is easy enough to do. However this then causes the index to fail, because when you add a new entry to the index it immediately sanity checks it by trying to retrieve it again - which fails because the comparison function can't find the thing it just added.

We can't really make changes in the database code itself because this is all in libcrypto and exposed via public API. We *could* just not create an index for the subject at all in the ca app. That would mean that every time we add an entry to the database we would have to scan each row in the database manually, checking to see if we have a duplicate. However that starts sounding like major heart surgery in a stable branch, which I'm a little reluctant to do.

While researching all of this I discovered a 16-year old bug introduced by commit 87e8feca. I suspect (but haven't checked) that this is actually when the crashing bug which started this saga off was also first introduced. It turns out that we have code in CA which is supposed to complain with a nice user friendly message if you attempt to add a duplicate entry into the database. It is supposed to tell you details of the clashing entry in the database. This doesn't work and you're just told that the operation failed with database error 2. Fixing that bug, also fixes the crashing bug co-incidentally! So, this means I'm going to remove the commit in this PR that actually adds the fix, and use this alternative fix instead.

**Comment by mattcaswell** [comment]
I've updated this PR with the new fix which also fixes the bug I described above. I've not done the unique subjects except when empty thing because of the problems also described above.

**Comment by mattcaswell** [comment]
I'll hold off updating the 1.0.2 version of this until we get some agreement that this is the way ahead.

**Comment by vdukhovni** [comment]
Given the hash table, if we were to allow non-unique subject names just for certs with an empty subject DN, the way to do that is to use the serial number as the hash key when the subject DN is empty.  The serial number really does need to be unique.

If that still too hard, I can live with the limitation that CAs that create empty subject DNs must set "unique_subject = no" (something that is generally sane anyway), but I if we can avoid the limitation with reasonable effort, then I'd prefer that.

The above basically means that empty subject DNs get implicitly mapped to the serial number for indexing, comparing, sorting, ...  Note that just arbitrarily comparing empty as non-equal can make some qsort implementations crash, since this can result in non-transitivity of the comparison function.  Don;'t know whether that's an issue here...

**Comment by mattcaswell** [comment]
> Given the hash table, if we were to allow non-unique subject names just for certs with an empty subject DN, the way to do that is to use the serial number as the hash key when the subject DN is empty. The serial number really does need to be unique.

Hmmm. That's a good idea. I will investigate if that is viable.

**Comment by mattcaswell** [comment]
It works! Pushed a new commit to do that. I think this is going to need some documentation somewhere. I've not added that yet. Waiting until we're all in agreement to this as an approach.

**Comment by levitte** [comment]
I'm a little lost...  it sounds like `unique_subject = no` never worked properly, is that a correct interpretation of what you wrote, @mattcaswell?

**Comment by mattcaswell** [comment]
> I'm a little lost... it sounds like unique_subject = no

No, the 16 year old bug I found is with "unique_subject = yes". If you attempt to add a duplicate subject you are supposed to get a nice error message. Instead you get "database error 2" (or something like that).

**Comment by levitte** [comment]
You misunderstand...

For empty subjects, you copy `row[DB_serial]` to `row[DB_name]` to force a unique name, right?  But what happens, then, with two rows having the name "/C=SE/CN=Richard Levitte" and `unique_subject = no`?  Your discussion above re hash tables and sorting requiring uniqueness suggests that it wouldn't work properly, and that has me worried.

I'll follow up with an in diff comment...

File: apps/ca.c [hunk 1]
```diff
@@ -1785,10 +1674,123 @@ static int do_body(X509 **xret, EVP_PKEY *pkey, X509 *x509,
...
+         * unique_subject is in use then we don't want different entries with
+         * empty subjects matching each other.
+         */
+        OPENSSL_free(row[DB_name]);
+        row[DB_name] = OPENSSL_strdup(row[DB_serial]);
+        if (row[DB_serial] == NULL) {
+            BIO_printf(bio_err, "Memory allocation failure\n");
             goto end;
+        }
+    }
```
Comment (levitte) [inline apps/ca.c]: The block above is an elaborate way to assume `unique_subject = no` at all times, and I wonder if it wouldn't be simpler to just...

File: apps/ca.c [hunk 2]
```diff
@@ -1785,10 +1674,123 @@ static int do_body(X509 **xret, EVP_PKEY *pkey, X509 *x509,
...
+
+    if (db->attributes.unique_subject) {
```
Comment (levitte) [inline apps/ca.c]: Change this condition to `(row[DB_name][0] != '\0' && db->attributes.unique_subject)`

**Comment by levitte** [comment]
I guess that in the end, I'm asking what would happen if we simply ignored `unique_subject` and switched to index by serial number (which is what `unique_subject = no` does for you)?  Considering a serial number within one CA database *must be unique anyway*, I have a very hard time seeing how this would cause a failure.

**Comment by mattcaswell** [comment]
> For empty subjects, you copy row[DB_serial] to row[DB_name] to force a unique name, right? But what happens, then, with two rows having the name "/C=SE/CN=Richard Levitte" and unique_subject = no? Your discussion above re hash tables and sorting requiring uniqueness suggests that it wouldn't work properly, and that has me worried.

No, the problem only occurs with unique_subject = yes, because in that case we create an index. If unique_subject = no then we do not create an index so there is no problem. Our issue was we wanted somewhere in the middle - a partial index - where most entries were unique, but some were not. Viktor's workaround of using serial number for those entries with an empty Subject works around that problem though.

File: apps/ca.c (hunk 1 above)
Comment (mattcaswell) [reply apps/ca.c]: > The block above is an elaborate way to assume unique_subject = no at all times

No? This is to explicitly handle the unique_subject = yes case to make sure that we really do have a unique subject, by filling in the Subject with a serial number iff the Subject is empty.

File: apps/ca.c (hunk 2 above)
Comment (mattcaswell) [reply apps/ca.c]: This doesn't solve the problem. The call to TXT_DB_insert() will fail later if we don't do the above block due to the existence of the index.

**Comment by levitte** [comment]
> Our issue was we wanted somewhere in the middle - a partial index - where most entries were unique, but some were not.

Ah, right, yeah I see the problem then.

Now bear with me...  what do you think would happen if we did a more radical change, such as throwing away the index on `DB_name` (i.e. effectively hard coding `unique_subject = no`)?  I'm trying to see if that would cause a failure, but cannot see how.  Do we foresee a performance hit?

**Comment by vdukhovni** [comment]
The problem is that users who rely on creation of duplicate certificates (by DN) to fail, will no longer have that happen.  Why they would want this is less clear, but perhaps some do.

I think that this PR should just address the issue, and removing "unique_subject" support could be a separate change?

**Comment by levitte** [comment]
Good point.  Sorry for derailing

**Comment by richsalz** [comment]
What's the status of this?

**Comment by mattcaswell** [comment]
> What's the status of this?

I ran into some problems with this PR. Needs some attention, but is currently on hold while I fight some bigger fires. I'll come back to it.

**Comment by mattcaswell** [comment]
I fixed a memory leak in this PR (discovered while backporting to 1.0.2).

@levitte please can you reconfirm, and also review #5445

**Comment by mattcaswell** [comment]
Grrr....it no longer cherry-picks to 1.1.0. See #5627.

**Comment by mattcaswell** [comment]
Pushed. Thanks!

**Pull request closed without merging**